### ISSUE TITLE
net7 - Build freezes when publishing trimmed app with net6+ class libraries

### ISSUE THREAD

**Issue by michaldobrodenka** [dotnet/runtime#87220]
I have a sln which contains several class libraries and one executable cmdline csproj. I'm modernizing whole solution, removing mono/net461 executable, upgrading class libraries from netstandard20 to net7.
When publishing my app with parameters:

`dotnet publish my.csproj -r linux-arm -p:PublishTrimmed=True -p:TrimMode=partial  --self-contained true -c Release -f net7.0 `

and one specific csproj of my class libraries is set to net7.0 build freezes. on line 
`C:\Program Files\dotnet\dotnet.exe exec "C:\Program Files\dotnet\sdk\7.0.101\Roslyn\bincore\csc.dll" ... `
Freezing means than NET host process takes almost all cpu, a several Gigs of RAM.

To "unfreeze" build I have to:
- remove trimming ( `-p:PublishTrimmed=True -p:TrimMode=partial` ) - without trimming build is completed in 20sec
- set this one library to netstandard20 and build with net7 SDK
- set all csproj files in solution to net6 and build in net6 SDK - net6 build completes in about a minute without problems

Building it on Windows 10,11 and in linux docker. 

What I've tried and didn't help:
- build with MS net7 SDK docker
- build for different runtimes - linux-x64, win-x64, linux-arm...
- SDK 7.0.100 and 7.0.101

Unfortunately project is closed source. But I can run some test, access my desktop remotely or something like that.

**Comment by michaldobrodenka** [comment]
@sbomer I've tried to build it with latest nighly, but seems that I can't get versions or something right:
![image](https://github.com/dotnet/runtime/assets/2361931/2006f83e-98b8-49f0-b418-22511e149413)

Maybe I'll wait for next official net8 Preview/net7 and retest it.

**Comment by michaldobrodenka** [comment]
```.NET SDK:
 Version:   8.0.100-preview.6.23307.25
 Commit:    f98e016a22

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100-preview.6.23307.25\

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.0-preview.6.23307.4
  Architecture: x64
  Commit:       54dab738de

.NET SDKs installed:
  2.1.800 [C:\Program Files\dotnet\sdk]
  6.0.402 [C:\Program Files\dotnet\sdk]
  7.0.200 [C:\Program Files\dotnet\sdk]
  7.0.400-preview.23225.8 [C:\Program Files\dotnet\sdk]
  8.0.100-preview.6.23307.25 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-preview.6.23307.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-preview.6.23307.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0-preview.6.23307.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

**Comment by michaldobrodenka** [comment]
Ok, after clearing nuget local cache, I'm able to start build, still freezes, but there is a problem with ILLink task version

![image](https://github.com/dotnet/runtime/assets/2361931/d282e7b7-1371-4386-a04c-f2cecabd5cf4)

```warning NU1603: TapHome.Core.Translations depends on Microsoft.NET.ILLink.Tasks (>= 7.0.100-1.23211.1) but Microsoft.NET.ILLink.Tasks 7.0.100-1.23211.1 was no
t found. An approximate best match of Microsoft.NET.ILLink.Tasks 8.0.0-preview.2.23128.3 was resolved.```

**Comment by michaldobrodenka** [comment]
I tried targeting net8 & latest net8 nightly (Preview6), build is still freezing. I'll retest it within a few weeks

**Comment by michaldobrodenka** [comment]
7.0.9 - still freezes
`<EnableTrimAnalyzer>false</EnableTrimAnalyzer>` - to one of the libraries and it builds ok!

**Comment by agocke** [comment]
This should be fixed. Newly discovered problems are likely new issues.

**Comment by michaldobrodenka** [comment]
> This should be fixed. Newly discovered problems are likely new issues.

Is the fix merged in 7.0.10? Because I tried it and build still hangs without `<EnableTrimAnalyzer>false</EnableTrimAnalyzer>` in one of the libraries


```
.NET SDK:
 Version:   7.0.400
 Commit:    73bf45718d

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.400\

Host:
  Version:      7.0.10
  Architecture: x64
  Commit:       a6dbb800a4

.NET SDKs installed:
  6.0.400 [C:\Program Files\dotnet\sdk]
  7.0.400-preview.23330.10 [C:\Program Files\dotnet\sdk]
  7.0.400 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.19 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  arm64 [C:\Program Files\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\arm64\InstallLocation]
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```
